src/helpers.rs

```diff
@@ -998,8 +998,12 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
     where
         &'a [OpTy<'tcx>; N]: TryFrom<&'a [OpTy<'tcx>]>,
     {
+        if abi.c_variadic {
+            throw_ub_format!(
+                "calling a non-variadic function with a variadic caller-side signature"
+            );
+        }
         self.check_abi_and_shim_symbol_clash(abi, exp_abi, link_name)?;
```

Comment:
You also reordered the check here. It probably makes sense to still check the other things first as they are more "fundamental" in a sense. So please keep the order you had in your first version: `check_abi_and_shim_symbol_clash`, then the variadic check, then the argument count.